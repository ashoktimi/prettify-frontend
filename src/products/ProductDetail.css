.ProductDetail{
    left: 0;
    right: 0;
    margin:auto;
    display: flex;
    justify-content: center;
}