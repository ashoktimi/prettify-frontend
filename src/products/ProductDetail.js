import React, { useContext } from 'react';
import { useDispatch } from "react-redux";
import UserContext from "../auth/userContext";
import { addToCart,  } from "../cart/cartSlice";
import defaultImage from '../images/spinning image.png'
import ProductRating from './ProductRating';
import { Container, Row, Col, Image, Button } from 'react-bootstrap';

const ProductDetail = ({id, name, price, priceSign, prevPrice, description, image_link, rating, numOfRating, color }) => {
  const product = {
    id,
    name,
    price,
    priceSign,
    prevPrice,
    image_link,
    rating,
    numOfRating,
  };
  const { handleToggleOffcanvas } = useContext(UserContext);
  const dispatch = useDispatch();
  const handleAddToCart = (product) => {
    dispatch(addToCart(product));
    handleToggleOffcanvas();
  };
  const imageUrl = image_link || defaultImage;

  return (  
      <Container style={{marginTop: '5rem'}}>
        <Row className="justify-content-center align-items-center">
          <Col xs={12} md={6} className="text-center mb-4 mb-md-0">
            <Image
              className="ProductDetail-image"
              src={imageUrl}
              alt={name}
              onError={(e) => {
                e.target.src = defaultImage;
              }}
              style={{width: '30rem', height: '30rem', backgroundColor: 'white'}}
            />
          </Col>
          <Col xs={12} md={6}>
            <div className="ProductDetail-info">
              <h5 className="ProductDetail-name">{name}</h5>
              <p className="ProductDetail-description">{description}</p>
                <small className="ProductCard-price"> {priceSign}{price} </small> 
                <small className="ms-2 text-muted priceBefore"> before {priceSign}{prevPrice} </small>
                <div className="ProductDetail-rating">
                  <small>({rating})</small>
                  <ProductRating rating={rating} />
                  <p>({numOfRating}) Reviews</p>   
                  {color && (
                    <div>
                      {color.map(c => (
                        <div
                          key={c.id}
                          style={{
                            backgroundColor: `${c.hex_value}`,
                            width: '30px',
                            height: '30px',
                            borderRadius: '50%',
                            display: 'inline-block',
                            marginLeft: '5px'
                          }}
                          />
                      ))}
                    </div>
                  )}       
                </div>          
                <Button variant="warning"  onClick={() => handleAddToCart(product)} style={{marginTop:"2rem"}} >ADD To CART</Button>
            </div>
          </Col>
        </Row>
      </Container>
    )
  }

export default ProductDetail;