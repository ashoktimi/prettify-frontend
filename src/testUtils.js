import React from "react";
import UserContext from "./auth/userContext";


const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  photo_url: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, handleToggleOffcanvas = () => false }) => (
    <UserContext.Provider value={{ currentUser, handleToggleOffcanvas }}>
      {children}
    </UserContext.Provider>
);

function setup(){
  window.matchMedia = jest.fn().mockImplementation(query => {
    return {
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    };
  });
}

function teardown(){
   window.matchMedia.mockClear();
   delete window.matchMedia;
}

function commonBeforeEach() {
  setup();
}

function commonAfterEach() {
  teardown();
}

export { UserProvider, commonBeforeEach, commonAfterEach };