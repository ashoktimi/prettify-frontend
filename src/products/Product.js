import React, { useState, useEffect } from "react";
import { useParams, useNavigate   } from "react-router-dom";
import PrettifyApi from "../api/api";
import LoadingSpinner from "../helpers/LoadingSpinner";
import ProductDetail from "./ProductDetail"

const Product = () =>{
    const { id } = useParams();
    const navigate = useNavigate();
    const [product, setProduct] = useState ({});

    useEffect(() => {
        async function getProduct (id){
          try {
              let data = await PrettifyApi.getProduct(id);
              setProduct(data);
            } catch (error) {
              console.log(error);
              alert(`${id} not found.`)
              navigate('/products')
            }
        }
        getProduct(id)
      }, [id, navigate]);

      if (!product) return <LoadingSpinner />;
      return(
          <div>
            {product
            ?(<div className="Product-div" key={product.id}>              
               <ProductDetail 
                id={product.id}
                name={product.name} 
                price={parseFloat(product.price).toFixed(2)}
                priceSign={product.price_sign}
                prevPrice={parseFloat(product.prev_price).toFixed(2)}
                description={product.description}
                image_link={product.image_link}
                rating={parseFloat(product.rating).toFixed(1)}
                numOfRating={product.number_rating}
                color={product.colors}
              />                
            </div>
            )
            :(
             <div>
                <p>{id} not found.</p>
             </div>
            )}
          </div>
      )
}

export default  Product;

