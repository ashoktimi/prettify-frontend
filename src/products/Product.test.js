import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { UserProvider,  commonBeforeEach, commonAfterEach } from "../testUtils";
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from '../cart/cartSlice';
import Product from "./Product";
import ProductDetail from "./ProductDetail";
import LoadingSpinner from "../helpers/LoadingSpinner";

beforeEach(commonBeforeEach)
afterEach(commonAfterEach)

describe('Product', () => {
 const store = configureStore({
   reducer: {
     cart: cartReducer
   }
 });

 test("renders without crashing", function () {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <UserProvider>
            <Product />
          </UserProvider>          
        </Provider>    
      </MemoryRouter>   
     );
  });
  
  test("matches snapshot", function () {
    const { asFragment } = 
     render(
      <MemoryRouter>
        <Provider store={store}>
         <UserProvider>
           <Product />
         </UserProvider>          
        </Provider>    
      </MemoryRouter>  
     );
    expect(asFragment()).toMatchSnapshot();
  });
  
  test("ProductDetail component renders correctly", function () {
    render(
      <MemoryRouter>
       <Provider store={store}>
        <UserProvider>
          <ProductDetail /> 
        </UserProvider> 
       </Provider>     
      </MemoryRouter>   
     );
  });
  
  test("LoadingSpinner component renders correctly ", function () {
    render(
      <MemoryRouter>
         <LoadingSpinner />
      </MemoryRouter>   
     );
  });
})
