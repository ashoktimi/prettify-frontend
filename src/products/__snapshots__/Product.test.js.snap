// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="Product-div"
    >
      <div
        class="container"
        style="margin-top: 5rem;"
      >
        <div
          class="justify-content-center align-items-center row"
        >
          <div
            class="text-center mb-4 mb-md-0 col-md-6 col-12"
          >
            <img
              class="ProductDetail-image"
              src="spinning image.png"
              style="width: 30rem; height: 30rem; background-color: white;"
            />
          </div>
          <div
            class="col-md-6 col-12"
          >
            <div
              class="ProductDetail-info"
            >
              <h5
                class="ProductDetail-name"
              />
              <p
                class="ProductDetail-description"
              />
              <small
                class="ProductCard-price"
              >
                 NaN 
              </small>
              <small
                class="ms-2 text-muted priceBefore"
              >
                 before NaN 
              </small>
              <div
                class="ProductDetail-rating"
              >
                <small>
                  (NaN)
                </small>
                <p>
                  () Reviews
                </p>
              </div>
              <button
                class="btn btn-warning"
                style="margin-top: 2rem;"
                type="button"
              >
                ADD To CART
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
